drivers/sds011/sds011.c

```diff
@@ -0,0 +1,351 @@
+{
+    assert(params->uart < UART_NUMOF);
+
+    if ((params->pwr_pin != GPIO_UNDEF) &&
+        (gpio_init(params->pwr_pin, GPIO_OUT) != 0)) {
+        return SDS011_ERROR;
+    }
+
+    memcpy(&dev->params, params, sizeof(sds011_params_t));
+
+    mutex_init(&dev->dev_lock);
+    mutex_init(&dev->cb_lock);
+
+    dev->cb = NULL;
+
+    sds011_power_on(dev);
+
+    return SDS011_OK;
+}
+
+int sds011_register_callback(sds011_t *dev, sds011_callback_t cb, void *ctx)
+{
+    mutex_lock(&dev->dev_lock);
+    dev->cbctx = ctx;
+    dev->cb = cb;
+
+    /* either register un unregister the uart callback */
+    if (uart_init(dev->params.uart, SDS011_UART_BAUDRATE,
+                  cb == NULL ? NULL : _rx_cb,
+                  cb == NULL ? NULL : dev) != 0) {
+        mutex_unlock(&dev->dev_lock);
+        return SDS011_ERROR;
+    }
+    mutex_unlock(&dev->dev_lock);
+    return SDS011_OK;
+}
+
+void sds011_power_on(const sds011_t *dev)
+{
+    if(dev->params.pwr_pin != GPIO_UNDEF) {
+        gpio_write(dev->params.pwr_pin, dev->params.pwr_ah);
+    }
+}
+
+void sds011_power_off(const sds011_t *dev)
+{
+    if(dev->params.pwr_pin != GPIO_UNDEF) {
+        gpio_write(dev->params.pwr_pin, !dev->params.pwr_ah);
+    }
+}
+
+int sds011_get_reporting_mode(sds011_t *dev, sds011_reporting_mode_t *mode)
+{
+    uint8_t cmd[] = {SDS011_CMD_DB1_SET_DR_MODE, SDS011_CMD_OPT_QUERY};
+    return _get_property(dev, cmd, sizeof(cmd), mode, SDS011_DB3_IDX);
+}
+
+int sds011_set_reporting_mode(sds011_t *dev, sds011_reporting_mode_t mode)
+{
+    uint8_t cmd[] = {SDS011_CMD_DB1_SET_DR_MODE, SDS011_CMD_OPT_SET, mode};
+
+    uint8_t recv[SDS011_FRAME_RECV_LEN];
+    int res = _send_recv_cmd(dev, cmd, sizeof(cmd), recv);
+
+    if (res == SDS011_OK) {
+        if ((recv[SDS011_DB2_IDX] == SDS011_CMD_OPT_SET) &&
+            (recv[SDS011_DB3_IDX] == mode)) {
+            return SDS011_OK;
+        }
+        else{
+            return SDS011_ERROR;
+        }
+    }
+
+    return res;
+}
+
+int sds011_read(sds011_t *dev, sds011_data_t *data)
+{
+    uint8_t cmd[] = {SDS011_CMD_DB1_QUERY_DATA};
+
+    uint8_t recv[SDS011_FRAME_RECV_LEN];
+    int res = _send_recv_cmd(dev, cmd, sizeof(cmd), recv);
+
+    if (res == SDS011_OK) {
+        data->pm_2_5 = recv[SDS011_DB1_IDX] | (recv[SDS011_DB2_IDX] << 8);
+        data->pm_10  = recv[SDS011_DB3_IDX] | (recv[SDS011_DB4_IDX] << 8);
+    }
+
+    return res;
+}
+
+int sds011_set_dev_id(sds011_t *dev, uint16_t sens_dev_id)
+{
+    uint8_t cmd[13] = {0};
+    cmd[0]  = SDS011_CMD_DB1_SET_DEV_ID;
+    cmd[11] = (sens_dev_id >> 8) & 0xFF;
+    cmd[12] = sens_dev_id & 0xFF;
+
+    uint8_t recv[SDS011_FRAME_RECV_LEN];
+    int res = _send_recv_cmd(dev, cmd, sizeof(cmd), recv);
+
+    if ((res == SDS011_OK) &&
+        (recv[SDS011_DB5_IDX] == cmd[11]) &&
+        (recv[SDS011_DB6_IDX] == cmd[12])) {
+        return SDS011_OK;
+    }
+    else {
+        return SDS011_ERROR;
+    }
+
+    return res;
```

Comment:
will never be reached, suggest to remove the `else` part above and simply return `SDS011_ERROR` here

Reply:
done